File: app/src/main/java/com/example/chupacabra/game/MazeGame.kt
package com.example.chupacabra.game

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

class MazeGame {
    // Maze: 0 = path, 1 = wall
    val maze = arrayOf(
        arrayOf(1,1,1,1,1,1,1,1,1,1),
        arrayOf(1,0,0,0,1,0,0,0,0,1),
        arrayOf(1,0,1,0,1,0,1,1,0,1),
        arrayOf(1,0,1,0,0,0,1,0,0,1),
        arrayOf(1,0,1,1,1,0,1,0,1,1),
        arrayOf(1,0,0,0,1,0,0,0,0,1),
        arrayOf(1,1,1,0,1,1,1,1,0,1),
        arrayOf(1,0,0,0,0,0,0,1,0,1),
        arrayOf(1,0,1,1,1,1,0,1,0,1),
        arrayOf(1,1,1,1,1,1,1,1,1,1),
    )

    val cellSize = 40
    
    // Player position state
    var playerX by mutableStateOf(1)
        private set
    var playerY by mutableStateOf(1)
        private set

    // Game state
    var isPaused by mutableStateOf(false)
        private set
    
    var revealedPaths by mutableStateOf(false)
        private set

    fun movePlayer(direction: Direction) {
        if (isPaused) return
        
        var newX = playerX
        var newY = playerY
        
        when (direction) {
            Direction.UP -> newY--
            Direction.DOWN -> newY++
            Direction.LEFT -> newX--
            Direction.RIGHT -> newX++
        }
        
        // Check bounds and walls
        if (newX >= 0 && newX < maze[0].size && 
            newY >= 0 && newY < maze.size && 
            maze[newY][newX] == 0) {
            playerX = newX
            playerY = newY
        }
    }
    
    fun jump(direction: Direction) {
        if (isPaused) return
        
        var newX = playerX
        var newY = playerY
        
        // Jump 2 cells in the direction
        when (direction) {
            Direction.UP -> newY -= 2
            Direction.DOWN -> newY += 2
            Direction.LEFT -> newX -= 2
            Direction.RIGHT -> newX += 2
        }
        
        // Check bounds and if destination is a path
        if (newX >= 0 && newX < maze[0].size && 
            newY >= 0 && newY < maze.size && 
            maze[newY][newX] == 0) {
            playerX = newX
            playerY = newY
        }
    }
    
    fun breakWall(direction: Direction) {
        if (isPaused) return
        
        var wallX = playerX
        var wallY = playerY
        
        when (direction) {
            Direction.UP -> wallY--
            Direction.DOWN -> wallY++
            Direction.LEFT -> wallX--
            Direction.RIGHT -> wallX++
        }
        
        // Check bounds and if it's a wall
        if (wallX >= 0 && wallX < maze[0].size && 
            wallY >= 0 && wallY < maze.size && 
            maze[wallY][wallX] == 1) {
            // Can't modify the original array directly in Compose
            // In a real implementation, you'd use a mutable state for the maze
        }
    }
    
    fun teleport(x: Int, y: Int) {
        if (isPaused) return
        
        // Check bounds and if destination is a path
        if (x >= 0 && x < maze[0].size && 
            y >= 0 && y < maze.size && 
            maze[y][x] == 0) {
            playerX = x
            playerY = y
        }
    }
    
    fun toggleRevealPath() {
        revealedPaths = !revealedPaths
    }
    
    fun togglePause() {
        isPaused = !isPaused
    }
}

enum class Direction {
    UP, DOWN, LEFT, RIGHT
}

File: src/game/MazeGame.ts
export enum Direction {
  UP = 'UP',
  DOWN = 'DOWN',
  LEFT = 'LEFT',
  RIGHT = 'RIGHT'
}

export interface GameState {
  maze: number[][]
  playerX: number
  playerY: number
  isPaused: boolean
  revealedPaths: boolean
}

const MAX_SPEED = 0.08     // units per frame (~4px/frame at cellSize 50)
const ACCELERATION = 0.015 // how fast velocity ramps up
const DECELERATION = 0.80  // friction when no input (multiplier)

export class MazeGame {
  private maze: number[][]
  private playerX: number = 1
  private playerY: number = 1
  private velocityX: number = 0
  private velocityY: number = 0
  private isPaused: boolean = false
  private revealedPaths: boolean = false
  readonly speed: number = 16

  constructor(maze: number[][]) {
    this.maze = maze
    
  }

  getMaze(): number[][] {
    return this.maze
  }

  getState(): GameState {
    return {
      maze: this.maze,
      playerX: this.playerX,
      playerY: this.playerY,
      isPaused: this.isPaused,
      revealedPaths: this.revealedPaths
    }
  }

  moveByDelta(dx: number, dy: number): GameState {
    if (this.isPaused) return this.getState()

    if (dx === 0 && dy === 0) {
      // No input — decelerate
      this.velocityX *= DECELERATION
      this.velocityY *= DECELERATION
      // Stop updating if velocity is negligible
      if (Math.abs(this.velocityX) < 0.001 && Math.abs(this.velocityY) < 0.001) {
        this.velocityX = 0
        this.velocityY = 0
        return this.getState()
      }
    } else {
      // Accelerate toward input direction
      this.velocityX += dx * ACCELERATION
      this.velocityY += dy * ACCELERATION

      // Cap at max speed
      const speed = Math.sqrt(this.velocityX ** 2 + this.velocityY ** 2)
      if (speed > MAX_SPEED) {
        this.velocityX = (this.velocityX / speed) * MAX_SPEED
        this.velocityY = (this.velocityY / speed) * MAX_SPEED
      }
    }

    this.playerX += this.velocityX
    this.playerY += this.velocityY
    return this.getState()
  }

movePlayer(dx: number, dy: number): GameState {
  if (this.isPaused) return this.getState()

  const newX = this.playerX + dx
  const newY = this.playerY + dy

  this.playerX = newX
  this.playerY = newY

  return this.getState()
}


  jump(direction: Direction): GameState {
    if (this.isPaused) return this.getState()

    let newX = this.playerX
    let newY = this.playerY

    if (direction === Direction.UP) newY -= this.speed * 4
    if (direction === Direction.DOWN) newY += this.speed * 4
    if (direction === Direction.LEFT) newX -= this.speed * 4
    if (direction === Direction.RIGHT) newX += this.speed * 4

    this.playerX = newX
    this.playerY = newY

    return this.getState()
  }

  breakWall(): GameState {
    return this.getState()
  }

  teleport(x: number, y: number): GameState {
    if (this.isPaused) return this.getState()

    this.playerX = x
    this.playerY = y

    return this.getState()
  }

  toggleRevealPath(): GameState {
    this.revealedPaths = !this.revealedPaths
    return this.getState()
  }

  togglePause(): GameState {
    this.isPaused = !this.isPaused
    return this.getState()
  }

  reset(): GameState {
    this.playerX = 1
    this.playerY = 1
    this.isPaused = false
    this.revealedPaths = false
    this.maze = this.maze.map(r => r.slice())
    return this.getState()
  }
}
